### ISSUE TITLE
Intent To Implement: Video Analytics

### ISSUE THREAD

**Issue by aghassemi** [ampproject/amphtml#9902]
## Summary
The goal is for `amp-video` and also other video players in AMP to issue events for major states to be reported through [`amp-analytics`](https://www.ampproject.org/docs/reference/components/amp-analytics).

https://github.com/ampproject/amphtml/issues/1654 and https://github.com/ampproject/amphtml/issues/2802 have some background on the requirements.

## Example
```
<amp-analytics>
<script type="application/json">
{
  "requests": {
    "event": "https://foo.com/va?vid=${videoId}&sid=${sessionId}&pt={playedTotal}&ap={autoplay}"
  },
  "triggers": {
     "videoSession": {
        "on": "video-session",
        "selector": "amp-video",
        "request": "event",
        "spec": {
           "end-session-when-invisible": true,
           "exclude-autoplay": false
        }
      }
    }
  }
}
</script>
</amp-analytics>
```

## Common Variables
All Video Triggers will provide the following variables:

| Var              | Type    | Description                                                                                                                                                                                                                                                                           |
|------------------|---------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ${videoid}       | String  | Id of the video element                                                                                                                                                                                                                                                               |
| ${videoDuration} | Number  | Total duration of the video in seconds                                                                                                                                                                                                                                                |
| ${muted}         | Boolean | Whether video was muted at the time of trigger                                                                                                                                                                                                                                        |
| ${paused}        | Boolean | Whether video was paused at the time of trigger                                                                                                                                                                                                                                       |
| ${ended}         | Boolean | Whether video had ended at the time of trigger                                                                                                                                                                                                                                        |
| ${currentTime}   | Number  | Current playback time in seconds at the time of trigger                                                                                                                                                                                                                               |
| ${fullscreen}    | Boolean | Whether video was in fullscreen at the time of trigger                                                                                                                                                                                                                                |
| ${autoplay}      | Boolean | Whether video was in autoplay mode at the time of trigger.  **NOTE**: This can be different than presence of `autoplay` attribute since 1- platform/user may have disable autoplay, 2- User may have, interacted with video and therefore exiting the autoplay mode by tapping,the video.    
| ${width}       | Number  | Width of video in PX                                                                                                                                                                                                                                                         |
| ${height} | Number  | Height of video in PX                                                                                                                                                                                                                                              |

**??QUESTION??**: Should these be specific to video triggers or also be exposed as Url substitutions (to be used in other triggers) or both? Do we want devs to use these values in a Timer trigger for instance?

## Triggers
### video-play (P0)
Issued when video is played (either autoplay or explicit manual play). `${autoplay}` variable
can be used to differentiate.

### video-pause (P0)
Issued when video is paused (either autoplay pause or explicit manual pause). `{autoplay}` variable
can be used to differentiate.

### video-ended (P0)
Issued when video ends.

### video-session (P0)
A video session starts when a video is played and ends when video pauses, ends 
or goes out of the view (if configured in the spec). Autoplaying videos can be optionally
removed from this reporting. Seeking or buffering does not end a session.

**QUESTION**: How about unload/document swipes in viewer? We need to end session and ping analytics. Document swipes in viewer should be fine, but how about page unload? Do we/can we use `onbeforeunload` to do this?

#### Spec
| Key              | Value    | Description                                                                                                                                                                                                                                                                           |
|------------------|---------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
|  "end-session-when-invisible"      | Boolean (false default) | Whether video going out of viewport should consider an end to the session even if it is still playing.                                                                                                                                                                                                                                                         |
| "exclude-autoplay"  | Boolean (false default)   | Whether autoplaying videos should be excluded from this trigger.                                                                                                                                                                                                                                               |

#### Additional variables
| Var              | Type    | Description                                                                                                                                                                                                                                                                           |
|---------------------------------|---------|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ${playedRangesJson} | String  | `Array` of `{start, end}` objects for each continuous playback segment as a JSON string. Overlapping ranges are possible if user replays a section. Every scrub will generate a new entry in this array.                                                                                                                                                                                                                                                             |
| ${scrubRangesJson}| String  | `Array` of `{start, end}` objects for each scrub as a JSON string. Overlapping ranges are possible.                                                                                                                                                                                                                                                       |
| ${sessionid} | String  | A unique id for the session in the format if {videoid}-{auto-incrementing-session-number} (e.g. myvideo-1)                                                                                                                                                                                                                                                      |
| ${playedTotal}   | Number | Total playing time for the session in seconds, includes overlapping watch segments.                                                                                                                                                                                                                                           |

### video-seconds-played (P0 - unless there is immediate interest)
Issued every time `x` seconds of video is played. Overlapping plays are still considered as play time.
This is essentially a timer trigger that is paused when video is paused.

#### Spec
| Key              | Value    | Description                                                                                                                                                                                                                                                                           |
|------------------|---------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
|  "Interval"      | Number (5 default) | How many seconds of playtime needs to pass before this fires.                                                                                                                                                                                                                                                   |

### video-fullscreen-enter (P1)
Issued when video enters fullscreen mode.

### video-fullscreen-exit (P1)
Issued when video exists fullscreen mode.

### video-autoplay-interacted (P1)
Issued when user interacts with an autoplaying video.

**NOTE**: The following triggers are low-priority P2 features assuming `video-session` trigger handles most of the common use-cases.

### video-scrub (P2 - unless there is immediate interest)
Issued when user is seeking within the video.

#### Additional variables
| Var              | Type    | Description                                                                                                                                                                                                                                                                           |
|------------------|---------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ${scrubStart}     | Number  | Start time of the scrub in seconds                                                                                                                                                                                                                                                           |
| ${scrubEnd}     | String  | End time of the scrub in seconds                                                                                                                                                                                                                                                       |

### video-percent-played (P2 - unless there is immediate interest)
Issued every time `x%` of total video duration is watched. `playedTimeRanges` is used
to calculated percentage of total duration watched. 
Overlapping play ranges are only counted **once** for calculation.

| Key              | Value    | Description                                                                                                                                                                                                                                                                           |
|------------------|---------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
|  "Interval"      | Number (25% default) | What percentage of playtime needs to pass before this fires.                                                                                                                                                                                                                                                   |

## How about other video players?
Video analytics will be implemented based on the `video-interface` rather than individual elements.
`video-interface` will be extended to include events that we would need to implement analytics _**however**_:
1. Not every player implements `video-interface` (8 do) 
2. Not every implementation will initially include the new events needed for analytics, as video players expose the needed events, analytics would be enabled for them. Initial focus would be on `amp-video` followed by `amp-ima-video`, `amp-iframe-video` and maybe `amp-youtube`.

Many 3P players provide their own separate video analytics anyway making this lower priority.

### amp-ima-video
@shawnbuso for insight. Does IMA SDK provide any sort of analytics for the 
ads and/or video itself? 
If not: 
1. We need `amp-ima-video` to support the same analytics as `amp-video`
2. Determine what Ad related triggers we need to add in addition. Would adding
two new `video-ad-start` and `video-ad-end` triggers suffice? 

### amp-iframe-video
This component does not exist yet, but scheduled. As long as the embedded video inside
the `amp-iframe-video` fully implements the `video-interface`, all analytics triggers should work.

### amp-x-player
As discussed, lower priority since many have their own analytics and adoption of the video-interface and extended events needed for analytics is usually slow.

**Comment by avimehta** [comment]
Hi @aghassemi,

I won't be able to attend the design review. Here are some comments about the I2I:

1. This looks great! Thanks for putting the proposal together.
1. `spec` should probably be renamed to be more specific? `videoSpec`?
1. Are these going to be available in amp-pixel?  I guess not because selectors can't be specified in amp-pixel. Just wanted to confirm.
1. `${ended}/${paused}` could be combined into `videoState` or something?  
1. Providing variables in timer trigger is tricky because you will need the video id there. Maybe make the variables accept videoId? A default can be provided in case of video triggers.
1. `video-fullscreen-enter` and other items in the variable section seem like triggers. Maybe mark them so?

**Comment by shawnbuso** [comment]
As far as IMA is concerned, we report metrics on the video ad for first frame (impression), 25%, 50%, 75%, 100%, and some other things depending on the ad format (skipped, interact (for VPAID ads), viewability, etc.). This is all part of the VAST video ad spec and is handled by the SDK internally - my amp plugin doesn't do anything special to make this happen (other than provide the IMA SDK with a handle to the video element to track progress). If publishers are using the IMA SDK they most likely aren't looking for additional ad metric reporting outside of what the SDK already offers, but they might be interested in some kind of reporting for the content, which we could provide.

**Comment by rudygalfi** [comment]
>> Should these be specific to video triggers or also be exposed as Url substitutions?
I think this could be useful but would require thought around how to use the substitutions. It might we worth considering to report these as a JSON object that contains all of the video state (play/pause, time position, etc.) based on specifying a variable like this:

`VIDEO_STATE(selector, reported_attributes_list)`

**Comment by wassgha** [comment]
Not sure how useful it might be but we might wanna include analytics for minimization events when implemented if there's interest. Just something to keep in mind.

**Comment by 1987yama3** [comment]
This is looks great!

But I want to track "Video Impression" event. Because I want to know the play start rate for each video. It is calculated from the number of impressions and the number of play start.

Note that, If visibilitySpec can be used to measure the number of impressions for each video ID, there is no problem.

**Comment by rudygalfi** [comment]
@aghassemi Do you think this will be in the June H2 sprint, or do you want to wait a bit for feedback on the design?

**Comment by aghassemi** [comment]
@rudygalfi we are planning for June H2 Sprint. There are some good feedbacks on the design here and having an initial implementation soon for folks to try out would be good for iterating on the solution.

**Comment by rudygalfi** [comment]
@aghassemi @cvializ What's the plan for tracking analytics work in the sprint process? Do we want to define two milestones at the top of this issue and call the first one done, or handle in some other way?

**Comment by aghassemi** [comment]
@rudygalfi @cvializ I suggest keeping this open, finish the last mile this sprint (documentation, remaining p0s, bugs/feedbacks from early adopters). Then consider it "launched", move remaining p3 features into new issues and close this.

**Comment by rudygalfi** [comment]
@aghassemi @cvializ Works for me

**Comment by lannka** [comment]
@aghassemi any update on this item?

**Comment by aghassemi** [comment]
@lannka Majority of the features are already in production. Last piece has a PR open. Remaining item is documentation before we can close this issue and considered it fully launched.

**Comment by omeryounus** [comment]
Is there any possibility to configure video analytics with google tag manager?

**Comment by avimehta** [comment]
@omeryounus not that I currently know of.

**Comment by sanchezedgar** [comment]
I'm wondering about the **P1** and **P2** triggers. Were the remaining features moved into new issues? I could not find open issues tracking them anywhere else. Thank you!

**Comment by cvializ** [comment]
Unless they are documented in the [video anayltics docs](https://github.com/ampproject/amphtml/blob/master/extensions/amp-analytics/amp-video-analytics.md), they were not implemented. If you need any of those P1 and P2 features, please [feel free to open a new issue](https://github.com/ampproject/amphtml/issues/new) and we will prioritize it. Thanks!
